shuf: include all echo args, not just the last

BenWiederhake opened the pull request:
This PR fixes all `shuf` bugs around repeated arguments:
- GNU shuf considers all arguments to `--echo`.  uutils `shuf` currently only considers the very last `--echo`, and ignores all the others.
- GNU shuf accepts multiple head-counts (`-n`), effectively only considering the *lowest*. uutils `shuf` currently only considers the very last `--head-count`, and ignores all the others.
- GNU shuf rejects repeated input ranges and output files with a clear error message. uutils `shuf` currently only considers the very last element each, silently ignoring all the others.

This is clearly buggy, since command-line arguments should be ignored only in exceptional cases. This appears to be a regression after #3329.

In this PR I'm fixing these and adopt the behavior of GNU, which is in my eyes the only sane reaction to those scenarios. Of course, this PR also adds tests for that.

I'm entirely new to uutils, so please do point out any resources that might help me. (I've already read {DEVELOPMENT,CONTRIBUTING,CODE_OF_CONDUCT}.md.)

<details><summary>Multiple echos example</summary>

```
$ shuf -e a b -e c
c
b
a
$ shuf -e a b -e c
a
b
c
```

The following demonstrations are run with this diff applied, to see what's going on:

```diff
diff --git a/src/uu/shuf/src/shuf.rs b/src/uu/shuf/src/shuf.rs
--- a/src/uu/shuf/src/shuf.rs
+++ b/src/uu/shuf/src/shuf.rs
@@ -48,9 +48,10 @@ mod options {
 #[uucore::main]
 pub fn uumain(args: impl uucore::Args) -> UResult<()> {
     let matches = uu_app().try_get_matches_from(args)?;
+    println!("**>{:?}<**", matches);
 
     let mode = if let Some(args) = matches.get_many::<String>(options::ECHO) {
-        Mode::Echo(args.map(String::from).collect())
+        Mode::Echo(args.map(String::from).inspect(|x| println!(">{x}<")).collect())
     } else if let Some(range) = matches.get_one::<String>(options::INPUT_RANGE) {
         match parse_range(range) {
             Ok(m) => Mode::InputRange(m),
```

Behavior *before* this PR:
```console
$ cargo run -- shuf -e a b -e c
**>ArgMatches { valid_args: ["echo", "input-range", "head-count", "output", "random-source", "repeat", "zero-terminated", "file", "help", "version"], valid_subcommands: [], args: FlatMap { keys: ["echo", "repeat", "zero-terminated"], values: [MatchedArg { source: Some(CommandLine), indices: [5], type_id: Some(alloc::string::String), vals: [[AnyValue { inner: alloc::string::String }]], raw_vals: [["c"]], ignore_case: false }, MatchedArg { source: Some(DefaultValue), indices: [6], type_id: Some(bool), vals: [[AnyValue { inner: bool }]], raw_vals: [["false"]], ignore_case: false }, MatchedArg { source: Some(DefaultValue), indices: [7], type_id: Some(bool), vals: [[AnyValue { inner: bool }]], raw_vals: [["false"]], ignore_case: false }] }, subcommand: None }<**
>c<
c
```

Note that `shuf` isn't even aware that `a` and `b` were passed in.

Behavior *after* this PR:
```console
$ cargo run -- shuf -e a b -e c
**>ArgMatches { valid_args: ["echo", "input-range", "head-count", "output", "random-source", "repeat", "zero-terminated", "file", "help", "version"], valid_subcommands: [], args: FlatMap { keys: ["echo", "repeat", "zero-terminated"], values: [MatchedArg { source: Some(CommandLine), indices: [2, 3, 5], type_id: Some(alloc::string::String), vals: [[AnyValue { inner: alloc::string::String }, AnyValue { inner: alloc::string::String }], [AnyValue { inner: alloc::string::String }]], raw_vals: [["a", "b"], ["c"]], ignore_case: false }, MatchedArg { source: Some(DefaultValue), indices: [6], type_id: Some(bool), vals: [[AnyValue { inner: bool }]], raw_vals: [["false"]], ignore_case: false }, MatchedArg { source: Some(DefaultValue), indices: [7], type_id: Some(bool), vals: [[AnyValue { inner: bool }]], raw_vals: [["false"]], ignore_case: false }] }, subcommand: None }<**
>a<
>b<
>c<
b
c
a
```

</details>

BenWiederhake commented:
CI failed, and I see a similar ICE panic when I run pre-commit (specifically `cargo +nightly clippy --workspace --all-targets --all-features --`). I don't really know what to do, and hope that the problem will somehow go away soon.

Anyway, stable clippy (`cargo clippy --workspace --all-targets --all-features --`) is clean, so I expect no problems.

tertsdiepraam commented:
You can ignore both failures, they're known issues :)

cakebaker commented on tests/by-util/test_shuf.rs:
```diff
@@ -103,6 +124,74 @@ fn test_head_count() {
+        .filter(|x| !x.is_empty())
+        .map(|x| x.parse().unwrap())
+        .collect();
+    result_seq.sort_unstable();
```
Why do you sort the result here and in `test_head_count_multi_small_then_big`?

BenWiederhake replied:
Good catch, thanks! That was a thoughtless copy-paste, because the other tests do it, too. Now that I think about it: You're right, it's not necessary, because the length of the result and membership of each (unique) expected element is already checked anyway.

Removed it in the next push.

cakebaker commented on tests/by-util/test_shuf.rs:
```diff
@@ -198,6 +287,24 @@ fn test_shuf_invalid_input_range_three() {
+#[test]
+fn test_shuf_multiple_outputs() {
+    new_ucmd!()
+        .args(&["-i", "2-9", "-o", "file_a", "-o", "file_b"])
```
```suggestion
        .args(&["-o", "file_a", "-o", "file_b"])
```

BenWiederhake replied:
Thanks! Removed it in the next push.

tertsdiepraam commented on src/uu/shuf/src/shuf.rs:
```diff
@@ -120,7 +120,6 @@ pub fn uu_app() -> Command {
         .version(crate_version!())
         .override_usage(format_usage(USAGE))
         .infer_long_args(true)
-        .args_override_self(true)
```
This is correct for almost all arguments, but not for `-r` and `-z`, so to avoid regressing those, we should add `overrides_with` to those arguments. A test for those cases would be great too :)

BenWiederhake replied:
Oh, right, absolutely!

Story how this happened: I saw `ArgAction::SetTrue` and blindly assumed that this would behave identically to python's argparse, which accepts repeated arguments. I did attempt to check this assumption, but apparently forgot to test it with `-r` and `-z`. TIL.

Will be a new commit in the next push.

tertsdiepraam reviewed (commented):
Nice work! I just think there's one case that's not handled correctly by this.

BenWiederhake commented:
Changes since last push:
- Made the tests `test_head_count_multi_small_then_big`, `test_head_count_multi_big_then_small`, and `test_shuf_multiple_outputs` sharper. Thanks @cakebaker!
- No longer introduce a bug regarding multiple `-r` and `-z` arguments, and add tests for it. Thanks @tertsdiepraam!
- Rebased on current main.

cakebaker commented:
Good job, thanks!
